File: android/app/src/main/kotlin/com/example/ntt_data/ui/SeamlessCircularProgressIndicator.kt
package com.example.ntt_data.ui
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

@Composable
fun SeamlessCircularProgressIndicator(
    progress: Float,
    maxProgress: Float = 100f,
    age: Double,
    borderColor: Color,
    drawArcColor: Color,
    years:String="",
    modifier: Modifier = Modifier,
    size: Dp = 100.dp,
    baseStrokeWidth: Dp = 8.dp
) {
    val normalizedProgress = (progress / maxProgress).coerceIn(0f, 1f)

    val animatedProgress by animateFloatAsState(
        targetValue = normalizedProgress,
        animationSpec = tween(durationMillis = 800),
        label = "AnimatedProgress"
    )

    val animatedAge by animateFloatAsState(
        targetValue = (normalizedProgress * age).toFloat(),
        animationSpec = tween(durationMillis = 800),
        label = "AnimatedAge"
    )

    val dynamicStrokeWidth = (baseStrokeWidth.value + (animatedProgress * 8)).dp

    Box(
        modifier = modifier.size(size),
        contentAlignment = Alignment.Center
    ) {
        Canvas(modifier = Modifier.fillMaxSize()) {
            val sweep = 360 * animatedProgress

            // Background arc
            drawArc(
                color = borderColor,
                startAngle = 0f,
                sweepAngle = 360f,
                useCenter = false,
                style = Stroke(width = (dynamicStrokeWidth.toPx() - 2), cap = StrokeCap.Butt)
            )

            // Foreground progress arc
            drawArc(
                color =drawArcColor ,
                startAngle = -90f,
                sweepAngle = sweep,
                useCenter = false,
                style = Stroke(width = dynamicStrokeWidth.toPx(), cap = StrokeCap.Butt)
            )
        }

        // Centered animated age text
         Text(
            text = "${age.toInt()}",
            color = Color.Black,
            fontSize = 18.sp
        )



    }
}

File: lib/modules/views/health_data/widgets/seamless_circular_progressIndicator.dart
import 'package:flutter/material.dart';
import 'dart:math';

class SeamlessCircularProgressIndicator extends StatefulWidget {
  final int progress; // Current progress value
  final int maxProgress;
  final int age;
  final Color borderColor;
  final Color drawArcColor;
  final double size;
  final double baseStrokeWidth;

  const SeamlessCircularProgressIndicator({
    super.key,
    required this.progress,
    this.maxProgress = 100,
    required this.age,
    required this.borderColor,
    required this.drawArcColor,
    this.size = 100,
    this.baseStrokeWidth = 8,
  });

  @override
  State<SeamlessCircularProgressIndicator> createState() =>
      _SeamlessCircularProgressIndicatorState();
}

class _SeamlessCircularProgressIndicatorState
    extends State<SeamlessCircularProgressIndicator>
    with SingleTickerProviderStateMixin {
  late AnimationController _controller;
  late Animation<double> _progressAnimation;
  late Animation<double> _ageAnimation;

  @override
  void initState() {
    super.initState();

    final normalizedProgress = (widget.progress / widget.maxProgress).clamp(
      0.0,
      1.0,
    );

    _controller = AnimationController(
      duration: const Duration(milliseconds: 800),
      vsync: this,
    );

    _progressAnimation = Tween<double>(
      begin: 0,
      end: normalizedProgress,
    ).animate(CurvedAnimation(parent: _controller, curve: Curves.easeInOut));

    _ageAnimation = Tween<double>(
      begin: 0,
      end: normalizedProgress * widget.age,
    ).animate(CurvedAnimation(parent: _controller, curve: Curves.easeInOut));

    _controller.forward();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: widget.size,
      height: widget.size,
      child: AnimatedBuilder(
        animation: _controller,
        builder: (context, child) {
          final strokeWidth =
              widget.baseStrokeWidth + (_progressAnimation.value * 8);
          return Stack(
            alignment: Alignment.center,
            children: [
              CustomPaint(
                size: Size(widget.size, widget.size),
                painter: _CirclePainter(
                  progress: _progressAnimation.value,
                  strokeWidth: strokeWidth,
                  borderColor: widget.borderColor,
                  drawArcColor: widget.drawArcColor,
                ),
              ),
              Text(
                "${widget.age.toInt()}",
                style: const TextStyle(fontSize: 18, color: Colors.black),
              ),
            ],
          );
        },
      ),
    );
  }
}

class _CirclePainter extends CustomPainter {
  final double progress;
  final double strokeWidth;
  final Color borderColor;
  final Color drawArcColor;

  _CirclePainter({
    required this.progress,
    required this.strokeWidth,
    required this.borderColor,
    required this.drawArcColor,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final rect = Offset.zero & size;
    final Paint backgroundPaint =
        Paint()
          ..color = borderColor
          ..strokeWidth = strokeWidth - 2
          ..style = PaintingStyle.stroke
          ..strokeCap = StrokeCap.butt;

    final Paint foregroundPaint =
        Paint()
          ..color = drawArcColor
          ..strokeWidth = strokeWidth
          ..style = PaintingStyle.stroke
          ..strokeCap = StrokeCap.butt;

    // Draw full circle background
    canvas.drawArc(rect, 0, 2 * pi, false, backgroundPaint);

    // Draw progress arc
    final sweepAngle = 2 * pi * progress;
    canvas.drawArc(rect, -pi / 2, sweepAngle, false, foregroundPaint);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => true;
}
